# **Title: Tracking Carbon Emission Trends Through Car Review Analysis using LLM**


**Project Description**

- Conducted carbon emission tracking within the entirety of the project's code and assessed its trend over time.
- Used a pre-trained LLM to classify the sentiment of the five car reviews in the car_reviews.csv dataset, and evaluated the classification accuracy and F1 score of predictions.
- Extracted and passed the first two sentences of the first review in the dataset to an English-to-Spanish translation LLM for Spanish customers. 
- Calculated the BLEU score to assess translation quality, using the content in reference_translations.txt as references.
- Loaded an extractive QA LLM such as "deepset/minilm-uncased-squad2" to formulate the question "What did he like about the brand?" and obtain an answer from 2nd Review to emphasize on brand aspects.
- Summarized the last review in the dataset, into approximately 50-55 tokens long. 

In [1]:
!pip install transformers
!pip install datasets
!pip install evaluate
!pip install codecarbon
from transformers import logging
logging.set_verbosity(logging.WARNING)

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 12.5 MB/s eta 0:00:00
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Defaulting to user installation because normal site-packages is not writeable
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 174.6/174.6 kB 18.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.1/53.1 kB 11.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 114.2 MB/s eta 0:00:00
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [2]:
import pandas as pd
import torch
import numpy as np
from codecarbon import EmissionsTracker

In [3]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Load the car reviews dataset
file_path = "data/car_reviews.csv"
df = pd.read_csv(file_path, delimiter=";")

# Put the car reviews and their associated sentiment labels in two lists
reviews = df['Review'].tolist()
real_labels = df['Class'].tolist()

emissions = tracker.stop()  


[codecarbon INFO @ 14:57:28] [setup] RAM Tracking...
[codecarbon INFO @ 14:57:28] [setup] GPU Tracking...
[codecarbon INFO @ 14:57:28] No GPU found.
[codecarbon INFO @ 14:57:28] [setup] CPU Tracking...
[codecarbon WARNING @ 14:57:28] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 14:57:29] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 14:57:29] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 14:57:29] >>> Tracker's metadata:
[codecarbon INFO @ 14:57:29]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 14:57:29]   Python version: 3.8.10
[codecarbon INFO @ 14:57:29]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 14:57:29]   Available RAM : 61.460 GB
[codecarbon INFO @ 14:57:29]   CPU count: 16
[codecarbon INFO @ 14:57:29]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 14:57:29]   GPU count: None
[code

In [30]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()

# Load a sentiment analysis LLM into a pipeline
from transformers import pipeline
# The model used is 'distilbert-base-uncased-finetuned-sst-2-english':
    # 'distilbert-base-uncased' refers to a distilled (simplified and faster) version of BERT that is case-insensitive.
    # 'finetuned-sst-2-english' indicates the model is fine-tuned on the SST-2 dataset, a benchmark for sentiment analysis,
    # specifically trained to detect sentiments in English text.
classifier = pipeline('sentiment-analysis', model='distilbert-base-uncased-finetuned-sst-2-english')

emissions = tracker.stop() 

[codecarbon INFO @ 15:04:42] [setup] RAM Tracking...
[codecarbon INFO @ 15:04:42] [setup] GPU Tracking...
[codecarbon INFO @ 15:04:42] No GPU found.
[codecarbon INFO @ 15:04:42] [setup] CPU Tracking...
[codecarbon WARNING @ 15:04:42] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:04:43] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:04:43] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:43] >>> Tracker's metadata:
[codecarbon INFO @ 15:04:43]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:04:43]   Python version: 3.8.10
[codecarbon INFO @ 15:04:43]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:04:43]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:04:43]   CPU count: 16
[codecarbon INFO @ 15:04:43]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:43]   GPU count: None
[code

In [31]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Perform inference on the car reviews and display prediction results
predicted_labels = classifier(reviews)
for review, prediction, label in zip(reviews, predicted_labels, real_labels):
    print(f"Review: {review}\nActual Sentiment: {label}\nPredicted Sentiment: {prediction['label']} (Confidence: {prediction['score']:.4f})\n")

emissions = tracker.stop() 

[codecarbon INFO @ 15:04:50] [setup] RAM Tracking...
[codecarbon INFO @ 15:04:50] [setup] GPU Tracking...
[codecarbon INFO @ 15:04:50] No GPU found.
[codecarbon INFO @ 15:04:50] [setup] CPU Tracking...
[codecarbon WARNING @ 15:04:50] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:04:51] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:04:51] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:51] >>> Tracker's metadata:
[codecarbon INFO @ 15:04:51]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:04:51]   Python version: 3.8.10
[codecarbon INFO @ 15:04:51]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:04:51]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:04:51]   CPU count: 16
[codecarbon INFO @ 15:04:51]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:51]   GPU count: None
[code

Review: I am very satisfied with my 2014 Nissan NV SL. I use this van for my business deliveries and personal use. Camping, road trips, etc. We dont have any children so I store most of the seats in my warehouse. I wanted the passenger van for the rear air conditioning. We drove our van from Florida to California for a Cross Country trip in 2014. We averaged about 18 mpg. We drove thru a lot of rain and It was a very comfortable and stable vehicle. The V8 Nissan Titan engine is a 500k mile engine. It has been tested many times by delivery and trucking companies. This is why Nissan gives you a 5 year or 100k mile bumper to bumper warranty. Many people are scared about driving this van because of its size. But with front and rear sonar sensors, large mirrors and the back up camera. It is easy to drive. The front and rear sensors also monitor the front and rear sides of the bumpers making it easier to park close to objects. Our Nissan NV is a Tow Monster. It pulls our 5000 pound travel tr

In [32]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Load accuracy and F1 score metrics    
import evaluate
accuracy = evaluate.load("accuracy")
f1 = evaluate.load("f1") # We use f1 score because it is useful for uneven classes and it considers both precision and recal
emissions = tracker.stop() 

[codecarbon INFO @ 15:04:58] [setup] RAM Tracking...
[codecarbon INFO @ 15:04:58] [setup] GPU Tracking...
[codecarbon INFO @ 15:04:58] No GPU found.
[codecarbon INFO @ 15:04:58] [setup] CPU Tracking...
[codecarbon WARNING @ 15:04:58] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:04:59] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:04:59] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:59] >>> Tracker's metadata:
[codecarbon INFO @ 15:04:59]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:04:59]   Python version: 3.8.10
[codecarbon INFO @ 15:04:59]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:04:59]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:04:59]   CPU count: 16
[codecarbon INFO @ 15:04:59]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:04:59]   GPU count: None
[code

In [33]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Map categorical sentiment labels into integer labels

references = [1 if label == "POSITIVE" else 0 for label in real_labels]
predictions = [1 if label['label'] == "POSITIVE" else 0 for label in predicted_labels]

emissions = tracker.stop() 

[codecarbon INFO @ 15:05:04] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:04] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:04] No GPU found.
[codecarbon INFO @ 15:05:04] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:04] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:06] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:06] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:06] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:06]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:06]   Python version: 3.8.10
[codecarbon INFO @ 15:05:06]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:06]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:06]   CPU count: 16
[codecarbon INFO @ 15:05:06]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:06]   GPU count: None
[code

In [34]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Calculate accuracy and F1 score
accuracy_result_dict = accuracy.compute(references=references, predictions=predictions)
accuracy_result = accuracy_result_dict['accuracy']
f1_result_dict = f1.compute(references=references, predictions=predictions)
f1_result = f1_result_dict['f1']
print(f"Accuracy: {accuracy_result}")
print(f"F1 result: {f1_result}")
emissions = tracker.stop() 



[codecarbon INFO @ 15:05:11] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:11] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:11] No GPU found.
[codecarbon INFO @ 15:05:11] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:11] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:12] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:12] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:12] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:12]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:12]   Python version: 3.8.10
[codecarbon INFO @ 15:05:12]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:12]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:12]   CPU count: 16
[codecarbon INFO @ 15:05:12]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:12]   GPU count: None
[code

Accuracy: 0.8
F1 result: 0.8571428571428571


Our sentiment analysis model, based on DistilBERT fine-tuned on SST-2, achieved an accuracy of 80% and an F1 score of approximately 0.857, demonstrating its effectiveness and accuracy in identifying the sentiment of car reviews. These results indicate that the model is reliable in classifying sentiments as either positive or negative, making it a valuable tool for analyzing customer feedback.








In [35]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()

# Load translation LLM into a pipeline and translate car review
first_review = reviews[0] #Extract the first review
translator = pipeline("translation", model="Helsinki-NLP/opus-mt-en-es") 

#This model utilizes the Transformer architecture, which is highly effective for translation tasks because it can handle long-range dependencies in text, making it suitable for complex sentence structures and idiomatic expressions. Using this model ensures that translations between English and Spanish are fluent and contextually appropriate.

translated_review = translator(first_review, max_length=27)[0]['translation_text']
print(f"Model translation:\n{translated_review}")

emissions = tracker.stop() 

[codecarbon INFO @ 15:05:16] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:16] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:16] No GPU found.
[codecarbon INFO @ 15:05:16] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:16] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:18] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:18] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:18] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:18]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:18]   Python version: 3.8.10
[codecarbon INFO @ 15:05:18]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:18]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:18]   CPU count: 16
[codecarbon INFO @ 15:05:18]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:18]   GPU count: None
[code

Model translation:
Estoy muy satisfecho con mi 2014 Nissan NV SL. Uso esta furgoneta para mis entregas de negocios y uso personal.


In [36]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Load reference translations from file
with open("data/reference_translations.txt", 'r') as file:
    lines = file.readlines()
references = [line.strip() for line in lines] #This strip function is used to clean up the lines, removing any leading and trailing whitespace, including the newline characters, making the data cleaner and easier to use in further processing.
print(f"Spanish translation references:\n{references}")
emissions = tracker.stop() 


[codecarbon INFO @ 15:05:27] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:27] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:27] No GPU found.
[codecarbon INFO @ 15:05:27] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:27] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:28] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:28] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:28] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:28]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:28]   Python version: 3.8.10
[codecarbon INFO @ 15:05:28]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:28]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:28]   CPU count: 16
[codecarbon INFO @ 15:05:28]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:28]   GPU count: None
[code

Spanish translation references:
['Estoy muy satisfecho con mi Nissan NV SL 2014. Utilizo esta camioneta para mis entregas comerciales y uso personal.', 'Estoy muy satisfecho con mi Nissan NV SL 2014. Uso esta furgoneta para mis entregas comerciales y uso personal.']


In [37]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
#BLEU (Bilingual Evaluation Understudy) is commonly used to evaluate the quality of machine-translated text against a set of reference translations.
bleu = evaluate.load("bleu")
# Load and calculate BLEU score metric
bleu = evaluate.load("bleu")
bleu_score = bleu.compute(predictions=[translated_review], references=[references])
print(bleu_score['bleu'])


emissions = tracker.stop() 

[codecarbon INFO @ 15:05:31] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:31] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:31] No GPU found.
[codecarbon INFO @ 15:05:31] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:31] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:32] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:32] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:32] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:32]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:32]   Python version: 3.8.10
[codecarbon INFO @ 15:05:32]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:32]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:32]   CPU count: 16
[codecarbon INFO @ 15:05:32]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:32]   GPU count: None
[code

0.6022774485691839


BLEU score ranges from 0 to 1 where 0.6023(60.23%) suggests that our translation has a moderate to good alignment with a refernce translations. It's not perfect but it conveys original meaning correctly but there maybe some difference in word coice , grammar or stylistic elements etc.

In [38]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()

# Import auto classes (optional: can be solved via pipelines too)
from transformers import AutoTokenizer
from transformers import AutoModelForQuestionAnswering
emissions = tracker.stop() 


[codecarbon INFO @ 15:05:36] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:36] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:36] No GPU found.
[codecarbon INFO @ 15:05:36] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:36] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:37] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:37] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:37] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:37]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:37]   Python version: 3.8.10
[codecarbon INFO @ 15:05:37]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:37]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:37]   CPU count: 16
[codecarbon INFO @ 15:05:37]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:37]   GPU count: None
[code

In [39]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()

# Instantiate model and tokenizer (For breaking down the text into tokens that the model can process)

model_ckp = "deepset/minilm-uncased-squad2"
tokenizer = AutoTokenizer.from_pretrained(model_ckp) 

# Load the question answering model from the same checkpoint. This model is specifically designed for the task of answering questions based on a given text context.

model = AutoModelForQuestionAnswering.from_pretrained(model_ckp)

emissions = tracker.stop() 

[codecarbon INFO @ 15:05:47] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:47] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:47] No GPU found.
[codecarbon INFO @ 15:05:47] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:47] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:49] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:49] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:49] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:49]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:49]   Python version: 3.8.10
[codecarbon INFO @ 15:05:49]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:49]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:49]   CPU count: 16
[codecarbon INFO @ 15:05:49]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:49]   GPU count: None
[code

In [40]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Define context and question, and tokenize them
context = reviews[1]
print(f"Context:\n{context}")
question = "What did he like about the brand?"
inputs = tokenizer(question, context, return_tensors="pt")

emissions = tracker.stop() 

[codecarbon INFO @ 15:05:55] [setup] RAM Tracking...
[codecarbon INFO @ 15:05:55] [setup] GPU Tracking...
[codecarbon INFO @ 15:05:55] No GPU found.
[codecarbon INFO @ 15:05:55] [setup] CPU Tracking...
[codecarbon WARNING @ 15:05:55] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:05:56] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:05:56] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:56] >>> Tracker's metadata:
[codecarbon INFO @ 15:05:56]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:05:56]   Python version: 3.8.10
[codecarbon INFO @ 15:05:56]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:05:56]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:05:56]   CPU count: 16
[codecarbon INFO @ 15:05:56]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:05:56]   GPU count: None
[code

Context:
The car is fine. It's a bit loud and not very powerful. On one hand, compared to its peers, the interior is well-built. The transmission failed a few years ago, and the dealer replaced it under warranty with no issues. Now, about 60k miles later, the transmission is failing again. It sounds like a truck, and the issues are well-documented. The dealer tells me it is normal, refusing to do anything to resolve the issue. After owning the car for 4 years, there are many other vehicles I would purchase over this one. Initially, I really liked what the brand is about: ride quality, reliability, etc. But I will not purchase another one. Despite these concerns, I must say, the level of comfort in the car has always been satisfactory, but not worth the rest of issues found.


In [42]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Perform inference and extract answer from raw outputs
with torch.no_grad():
    
#The model outputs the logits (scores) that represent the likelihood of each token being the start and end of the answer.
  outputs = model(**inputs) 
 
#Identify the position of the highest score in start_logits to determine where the answer starts and ends.
start_idx = torch.argmax(outputs.start_logits) 
end_idx = torch.argmax(outputs.end_logits) + 1 

# Extract the token IDs of the answer span based on the start and end indices.
answer_span = inputs["input_ids"][0][start_idx:end_idx]

#Decode the token IDs back into the text of the answer.
answer = tokenizer.decode(answer_span)
print("Answer: ", answer)
emissions = tracker.stop() 


[codecarbon INFO @ 15:06:07] [setup] RAM Tracking...
[codecarbon INFO @ 15:06:07] [setup] GPU Tracking...
[codecarbon INFO @ 15:06:07] No GPU found.
[codecarbon INFO @ 15:06:07] [setup] CPU Tracking...
[codecarbon WARNING @ 15:06:07] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:06:09] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:06:09] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:09] >>> Tracker's metadata:
[codecarbon INFO @ 15:06:09]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:06:09]   Python version: 3.8.10
[codecarbon INFO @ 15:06:09]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:06:09]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:06:09]   CPU count: 16
[codecarbon INFO @ 15:06:09]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:09]   GPU count: None
[code

Answer:  ride quality, reliability


It indicates that the reviewer values both the physical comfort and driving experience provided by the car, as well as its ability to operate without issues consistently

In [43]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Get original text to summarize upon car review 
text_to_summarize = reviews[-1] 
print(f"Original text:\n{text_to_summarize}")
emissions = tracker.stop() 


[codecarbon INFO @ 15:06:12] [setup] RAM Tracking...
[codecarbon INFO @ 15:06:12] [setup] GPU Tracking...
[codecarbon INFO @ 15:06:12] No GPU found.
[codecarbon INFO @ 15:06:12] [setup] CPU Tracking...
[codecarbon WARNING @ 15:06:12] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:06:13] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:06:13] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:13] >>> Tracker's metadata:
[codecarbon INFO @ 15:06:13]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:06:13]   Python version: 3.8.10
[codecarbon INFO @ 15:06:13]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:06:13]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:06:13]   CPU count: 16
[codecarbon INFO @ 15:06:13]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:13]   GPU count: None
[code

Original text:
I've been dreaming of owning an SUV for quite a while, but I've been driving cars that were already paid for during an extended period. I ultimately made the decision to transition to a brand-new car, which, of course, involved taking on new payments. However, given that I don't drive extensively, I was inclined to avoid a substantial financial commitment. The Nissan Rogue provides me with the desired SUV experience without burdening me with an exorbitant payment; the financial arrangement is quite reasonable. Handling and styling are great; I have hauled 12 bags of mulch in the back with the seats down and could have held more. I am VERY satisfied overall. I find myself needing to exercise extra caution when making lane changes, particularly owing to the blind spots resulting from the small side windows situated towards the rear of the vehicle. To address this concern, I am actively engaged in making adjustments to my mirrors and consciously reducing the frequency of la

In [44]:
# Initialize the emissions tracker
tracker = EmissionsTracker()
tracker.start()
# Load summarization pipeline and perform inference
#T5-small-booksum summarization model allows for leveraging advanced AI tools to automate the summarization process
#Ensures consistency and scalability in handling customer reviews
model_name = "cnicu/t5-small-booksum"
summarizer = pipeline("summarization", model=model_name)
outputs = summarizer(text_to_summarize, max_length=53)
summarized_text = outputs[0]['summary_text']
print(f"Summarized text:\n{summarized_text}")
emissions = tracker.stop() 

[codecarbon INFO @ 15:06:23] [setup] RAM Tracking...
[codecarbon INFO @ 15:06:23] [setup] GPU Tracking...
[codecarbon INFO @ 15:06:23] No GPU found.
[codecarbon INFO @ 15:06:23] [setup] CPU Tracking...
[codecarbon WARNING @ 15:06:23] No CPU tracking mode found. Falling back on CPU constant mode.
[codecarbon WARNING @ 15:06:24] We saw that you have a AMD EPYC 7R13 Processor but we don't know it. Please contact us.
[codecarbon INFO @ 15:06:24] CPU Model on constant consumption mode: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:24] >>> Tracker's metadata:
[codecarbon INFO @ 15:06:24]   Platform system: Linux-5.10.209-198.858.amzn2.x86_64-x86_64-with-glibc2.29
[codecarbon INFO @ 15:06:24]   Python version: 3.8.10
[codecarbon INFO @ 15:06:24]   CodeCarbon version: 2.3.5
[codecarbon INFO @ 15:06:24]   Available RAM : 61.460 GB
[codecarbon INFO @ 15:06:24]   CPU count: 16
[codecarbon INFO @ 15:06:24]   CPU model: AMD EPYC 7R13 Processor
[codecarbon INFO @ 15:06:24]   GPU count: None
[code

Summarized text:
the Nissan Rogue provides me with the desired SUV experience without burdening me with an exorbitant payment; the financial arrangement is quite reasonable. I have hauled 12 bags of mulch in the back with the seats down and could have held more.


In [45]:
# Load emissions data
data_path = 'emissions.csv'
df_Carbon = pd.read_csv(data_path)


In [46]:
print(df_Carbon.columns)


Index(['timestamp', 'project_name', 'run_id', 'duration', 'emissions',
       'emissions_rate', 'cpu_power', 'gpu_power', 'ram_power', 'cpu_energy',
       'gpu_energy', 'ram_energy', 'energy_consumed', 'country_name',
       'country_iso_code', 'region', 'cloud_provider', 'cloud_region', 'os',
       'python_version', 'codecarbon_version', 'cpu_count', 'cpu_model',
       'gpu_count', 'gpu_model', 'longitude', 'latitude', 'ram_total_size',
       'tracking_mode', 'on_cloud', 'pue'],
      dtype='object')


In [47]:
df_Carbon.head()

timestamp project_name  ... on_cloud  pue
0  2024-04-26T02:02:44   codecarbon  ...        N  1.0
1  2024-04-26T02:03:24   codecarbon  ...        N  1.0
2  2024-04-26T02:04:00   codecarbon  ...        N  1.0
3  2024-04-26T02:05:41   codecarbon  ...        N  1.0
4  2024-04-26T02:05:50   codecarbon  ...        N  1.0

[5 rows x 31 columns]

In [48]:
df_Carbon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   timestamp           120 non-null    object 
 1   project_name        120 non-null    object 
 2   run_id              120 non-null    object 
 3   duration            120 non-null    float64
 4   emissions           120 non-null    float64
 5   emissions_rate      120 non-null    float64
 6   cpu_power           120 non-null    float64
 7   gpu_power           120 non-null    float64
 8   ram_power           120 non-null    float64
 9   cpu_energy          120 non-null    float64
 10  gpu_energy          120 non-null    int64  
 11  ram_energy          120 non-null    float64
 12  energy_consumed     120 non-null    float64
 13  country_name        120 non-null    object 
 14  country_iso_code    120 non-null    object 
 15  region              120 non-null    object 
 16  cloud_pr

In [49]:
df_Carbon.describe()

duration     emissions  ...  ram_total_size    pue
count  120.000000  1.200000e+02  ...      120.000000  120.0
mean     4.440704  2.982432e-05  ...       61.460468    1.0
std     11.913206  8.003851e-05  ...        0.000000    0.0
min      0.002557  8.400040e-09  ...       61.460468    1.0
25%      0.007572  4.097508e-08  ...       61.460468    1.0
50%      0.396303  2.653528e-06  ...       61.460468    1.0
75%      4.075624  2.738191e-05  ...       61.460468    1.0
max     78.314163  5.263268e-04  ...       61.460468    1.0

[8 rows x 19 columns]

In [51]:
import plotly.express as px

# Assume df_Carbon is your DataFrame with the required data
fig = px.line(df_Carbon, x='timestamp', y='emissions', title='Carbon Emissions Over Time',
              labels={'timestamp': 'Time', 'emissions': 'CO2 Emissions (grams)'},
              template='plotly_dark')

# Update layout with x-axis title, y-axis title, and modify x-axis for range slider and buttons
fig.update_layout(
    xaxis_title='Time',
    yaxis_title='Carbon Emissions (grams)',
    xaxis=dict(
        rangeslider=dict(visible=True),
        rangeselector=dict(
            buttons=list([
                dict(count=30, label="30s", step="second", stepmode="backward"),
                dict(count=2, label="2m", step="minute", stepmode="backward"),
                dict(count=5, label="5m", step="minute", stepmode="backward"),
                dict(count=10, label="10m", step="minute", stepmode="backward"),
                dict(count=30, label="30m", step="minute", stepmode="backward"),
                dict(count=1, label="1h", step="hour", stepmode="backward"),
                dict(step="all")
            ])
        ),
        type='date'
    ),
    yaxis=dict(type='linear')
)

fig.show()
